### PR TITLE
Provide an Inference engine for ctranslate2

### PR DESCRIPTION
**Pull request by l-k-11235**
This PR provides an inference engine for ctranslate2 (parallel tensor mode not available for now).
It simplifies inference in the vicuna chat-bot using the -py and ctranslate2 inference engines.

### CODE REVIEW

File: onmt/inference_engine_ct2.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        opt: inference options
+    """
+
+    def __init__(self, opt):
+        self.opt = opt
+        self.transforms_cls = get_transforms_cls(self.opt._all_transform)
+        self.logger = logger
+        if opt.world_size == 1:
+            self.device_id = 0
+
```
Comment (vince62s) [inline onmt/inference_engine_ct2.py:24]: do we really need these warm_up, build_ct2_translator, build_tokenizer ?
I think we can stick to the inference_engine.py logic and put eveything in __init__() but if there is a good reason then fine.

File: onmt/inference_engine_ct2.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        if opt.model_task == ModelTask.LANGUAGE_MODEL:
+            translator = ctranslate2.Generator(
+                opt.models[0], device="cuda", device_index=opt.gpu_ranks
+            )
+        else:
+            translator = ctranslate2.Translator(
+                self.opt.models[0], device="cuda", device_index=opt.gpu_ranks
+            )
+        return translator
+
```
Comment (vince62s) [inline onmt/inference_engine_ct2.py:40]: not sure why _translate() and translate_batch()
is one of the two supposed to be exposed outside ?
if not then use _translate_batch()

File: onmt/inference_engine_ct2.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    def build_tokenizer(self):
+        if self.transforms_cls.get("sentencepiece", None) is not None:
+            from onmt.transforms.tokenize import SentencePieceTransform
+
+            self.tokenizer = SentencePieceTransform(self.opt)
+            self.tokenizer.warm_up()
+            self.build_sentencepiece_vocab()
+        else:
+            logger.error("Only the sentencepiece tokenize transform is supported.")
```
Comment (vince62s) [inline onmt/inference_engine_ct2.py:93]: is there a reason for this ?

File: onmt/inference_engine_ct2.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        scores = []
+        preds = []
+        for batch in infer_iter:
+            _scores, _preds = self.translate_batch(batch, opt, add_bos)
+            scores += _scores
+            preds += _preds
+        return scores, preds
+
+    def translate_batch(self, batch, opt, add_bos=True):
+        if opt.model_task == ModelTask.LANGUAGE_MODEL:
```
Comment (vince62s) [inline onmt/inference_engine_ct2.py:51]: else SEQ_2_SEQ ?

File: docs/source/examples/replicate_vicuna/chatbot.py [hunk 5]
```diff
@@ -1,32 +1,43 @@
...
+from onmt.transforms.tokenize import SentencePieceTransform
 from onmt.utils.parse import ArgumentParser
 from onmt.utils.misc import use_gpu, set_random_seed
 
-inf_type = "-py"
-# inf_type = "ct2"
+parser = argparse.ArgumentParser()
+parser.add_argument(
+    "--inference_config_file",
+    "-inference_config_file",
```
Comment (francoishernandez) [inline docs/source/examples/replicate_vicuna/chatbot.py:14]: Either do this "--"/"-" duality for all flags, or none, but doing it only for one is misleading.

File: docs/source/examples/replicate_vicuna/translate_opts_ct2.yaml [hunk 6]
```diff
@@ -0,0 +1,17 @@
+model_task: 'lm'
+model: finetuned_llama7B/llama7B-vicuna-onmt_step_4000.concat_CT2
+transforms: [sentencepiece]
+src_subword_model: llama/tokenizer.model
+tgt_subword_model: llama/tokenizer.model
+src_subword_vocab: finetuned_llama7B/llama7B-vicuna-onmt_step_4000.concat_CT2/vocabulary.json
+world_size: 1
+gpu_ranks: [0]
+max_length: 50
```
Comment (francoishernandez) [inline docs/source/examples/replicate_vicuna/translate_opts_ct2.yaml:9]: Is max_length supposed to be this low?

File: onmt/inference_engine_ct2.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+        preds = []
+        for batch in infer_iter:
+            _scores, _preds = self.translate_batch(batch, self.opt)
+            scores += _scores
+            preds += _preds
+        return scores, preds
+
+    def infer_file(self):
+        """File inference. Source file must be the opt.src argument"""
+        if self.opt.world_size == 1:
```
Comment (francoishernandez) [inline onmt/inference_engine_ct2.py:123]: else what happens?

File: onmt/inference_engine_ct2.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+                    for out in translated_batch
+                ],
+                [],
+            )
+            scores = sum([out.scores for out in translated_batch], [])
+
+        return scores, preds
+
+    def infer_list(self, src):
+        if self.opt.world_size == 1:
```
Comment (francoishernandez) [inline onmt/inference_engine_ct2.py:104]: else what happens?

**Review by francoishernandez — COMMENTED** [review]
Some comments.

File: docs/source/examples/replicate_vicuna/simple_inference.py [hunk 9]
```diff
@@ -22,21 +36,34 @@ def _get_parser():
...
     start = time.time()
     scores, preds = engine.infer_file()
-    scores = [_score.cpu().numpy().tolist() for _score in sum(scores, [])]
+    engine.terminate()
     dur = time.time() - start
     print(f"Time to generate {len(preds)} answers: {dur}s")
+    if inference_mode == "py":
+        scores = [
+            [_score.cpu().numpy().tolist() for _score in _scores] for _scores in scores
+        ]
```
Comment (francoishernandez) [inline docs/source/examples/replicate_vicuna/simple_inference.py:57]: Specific reason for not doing this in the `infer_file` method? This means the two subclasses don't have the exact same interface?

File: docs/source/examples/replicate_vicuna/translate_opts_py.yaml [hunk 10]
```diff
@@ -1,10 +1,11 @@
-model_task: 'lm'
-model: finetuned_llama7B/llama7B-vicuna-onmt_step_4000.concat.pt
+model_task: "lm"
+model: finetuned_llama7B / llama7B - vicuna - onmt_step_4000.concat.pt
 transforms: [sentencepiece]
-src_subword_model: llama/tokenizer.model
-tgt_subword_model: llama/tokenizer.model
+src_subword_model: llama / tokenizer.model
+tgt_subword_model: llama / tokenizer.model
```
Comment (francoishernandez) [inline docs/source/examples/replicate_vicuna/translate_opts_py.yaml:5]: Are these whitespaces expected?

**Review by francoishernandez — COMMENTED** [review]
Some small comments, else we're probably good.
Do we want to add a few tests? E.g. a simple integration test of simple_inference.py might not hurt.

File: docs/source/examples/replicate_vicuna/simple_inference.py (hunk 9 above)
Comment (l-k-11235) [reply docs/source/examples/replicate_vicuna/simple_inference.py:57]: Yes indeed, py scores are lists of tensors unlike ctranslate scores.

File: docs/source/examples/replicate_vicuna/translate_opts_py.yaml (hunk 10 above)
Comment (l-k-11235) [reply docs/source/examples/replicate_vicuna/translate_opts_py.yaml:5]: oh it seems that I applied black on the inference config file ...

File: docs/source/examples/replicate_vicuna/chatbot.py [hunk 11]
```diff
@@ -134,7 +136,7 @@ def make_bot_message(prompt, inference_mode):
         bot_message = "\n".join(sent[0] for sent in predictions)
     elif inference_mode == "ct2":
         scores, predictions = CACHE["inference_engine"].infer_list(src)
-        bot_message = "\n".join(sent for sent in predictions)
+        bot_message = "\n".join(sent[0] for sent in predictions)
```
[the remark this answers is not in the document]
Comment (l-k-11235) [reply docs/source/examples/replicate_vicuna/chatbot.py:139]: In this chat context, the hypotheses are lists of one element but we still need to take the first one.

**Pull request merged**